doc/dev/ragged_iter_domain_design_doc.md

```diff
@@ -0,0 +1,914 @@
+
+```cpp
+// User-defined Fusion
+Fusion fusion;
+FusionGuard fg(&fusion);
+
+// User provides data and offsets as separate inputs
+auto tv_data = TensorViewBuilder()
+    .ndims(2)
+    .shape({-1, 512})  // [total_tokens, hidden]
+    .dtype(DataType::Float)
+    .build();
+fusion.addInput(tv_data);
+
+auto tv_offsets = TensorViewBuilder()
+    .ndims(1)
+    .shape({9})  // [num_experts + 1]
+    .dtype(DataType::Int)
+    .build();
+fusion.addInput(tv_offsets);
+
+// User explicitly creates nested tensor
+auto tv_nested = asNested(tv_data, tv_offsets, /*ragged_dim=*/0);
+// tv_nested has shape [batch=8, ragged_tokens, hidden=512]
+
+// Operations on the nested tensor
+auto tv_result = some_operation(tv_nested);
+
+fusion.addOutput(tv_result);
+```
+
+The output tensor, `tv_result`, is a nested tensor. The extents of the nested domains are given as a fusion input, but in general, they are not known until the fusion is executed. Thus, if the nested tensor struct were defined like:
+
+```cpp
+template <typename DT, int rank>
+struct NestedTensor {
+	DT* ptr;
+	int64_t extents[rank];
+	int64_t nested_domain_extents[ragged_dimension_rank];
+};
+```
+
+The value of `nested_domain_extents` is not available until the completion of the kernel, which would block the launch of the subsequent kernel.
+
+Instead, we would like the fusion to be defined as follows:
+
+```cpp
+fusion.addInput(tv_data);      // Original data input (unchanged)
+fusion.addInput(tv_offsets);   // Original offset input (unchanged)
+
+auto tv_nested = asNested(tv_data, tv_offsets, /*ragged_dim=*/0);
+auto tv_result = some_operation(tv_nested);
+
+auto tv_result_offsets = /* extract/compute offset part of tv_result */;
+
+fusion.addOutput(tv_result);      // Data tensor output
+fusion.addOutput(tv_result_offsets);   // Offset tensor output (injected)
+```
+
+Here, for `tv_result` we would use the same `Tensor` struct as the normal tensor. The offset tensor would be a 1D tensor with the `ptr` val referring to the vector holding the offsets on the device memory. In this case, there's nothing to block the launch of the subsequent kernel as the offset vector would remain on the device memory.
+
+Since it is an implementation detail, the offset tensor should be hidden behind the nested tensor in the user-facing Fusion definition. When a user uses `asNested` to create a nested tensor, it should still create a single nested tensor Val, as illustrated in the first case above. The translation to the second pattern should be done automatically, e.g., by a new preseg pass.
+
+#### Runtime Binding (Initial Version)
+
+**Challenge**: How should `ExpressionEvaluator` bind a ragged TensorView to actual tensor objects at runtime?
+
+**Initial Approach**: In this initial version, we take a simplified approach that avoids the complexity of custom tensor types:
+
+1. **Input Nested Tensors**: Must be passed as **two separate parameters**:
+   - A normal `at::Tensor` containing the flattened data
```

Comment:
Can you write some examples? The word "flattened" is slightly overloaded. I don't think you actually mean a 1D tensor here for things like `[e, d, j1, h]`.

Reply:
So by "flattened" you do mean **1D**? That's fine for the first version, but I'm sure you understand that won't work with GroupedMmaOp or EP?

Reply:
No. It just merges back the partitioned sub iterdomains back to a standard iter domain. TensorDomain can be multidimensional.